secure/storage/src/tests/vault.rs

```diff
@@ -6,42 +6,174 @@ use crate::{
+/// strict separation between unique namespaces. This test depends on running Vault, which can be
+/// done by using the provided docker run script in `docker/vault/run.sh`
 #[test]
 #[ignore]
 fn execute_storage_tests_vault() {
-    let mut storage = VaultStorage::new(VAULT_HOST.into(), VAULT_ROOT_TOKEN.into(), None);
-    storage.reset_and_clear().unwrap();
+    let mut storage = create_initialized_vault_storage(None);
+    storage
+        .reset_and_clear()
+        .expect("Failed to reset VaultStorage after creation!");
 
-    test_vault(&mut storage);
-    suite::execute_all_storage_tests(&mut storage);
+    for test in VAULT_TESTS.iter() {
+        test();
+        storage
+            .reset_and_clear()
+            .expect("Failed to reset storage engine between tests!");
+    }
+}
+
+/// Verifies that when calling reset on a VaultStorage instance, if the instance was created with
+/// a namespace, only the secrets within the namespace are removed. Otherwise, the entire storage
+/// instance is reset. This helps to ensure operations across namespaces do not interfere.
+fn test_vault_namespace_reset() {
+    // Create two distinct VaultStorage namespaces with difference pieces of data
+    let mut storage_1 = create_initialized_vault_storage(Some(VAULT_NAMESPACE_1.into()));
+    let mut storage_2 = create_initialized_vault_storage(Some(VAULT_NAMESPACE_2.into()));
+    storage_1
+        .create(U64_KEY_1, Value::U64(U64_VALUE_1), &Policy::public())
+        .unwrap();
+    storage_2
+        .create(U64_KEY_2, Value::U64(U64_VALUE_2), &Policy::public())
+        .unwrap();
+
+    // Verify stores/resets do not interfere across namespaces
+    storage_1.reset_and_clear().expect(&format!(
+        "Failed to reset VaultStorage with namespace: {}!",
+        VAULT_NAMESPACE_1
+    ));
+    assert!(storage_1.get(U64_KEY_1).is_err());
+    assert_eq!(
+        storage_2.get(U64_KEY_2).unwrap().value.u64().unwrap(),
+        U64_VALUE_2
+    );
+
+    storage_1
+        .create(U64_KEY_1, Value::U64(U64_VALUE_1), &Policy::public())
+        .unwrap();
+    storage_2.reset_and_clear().expect(&format!(
+        "Failed to reset VaultStorage with namespace: {}!",
+        VAULT_NAMESPACE_2
+    ));
+    assert_eq!(
+        storage_1.get(U64_KEY_1).unwrap().value.u64().unwrap(),
+        U64_VALUE_1
+    );
+    storage_2
+        .create(U64_KEY_2, Value::U64(U64_VALUE_2), &Policy::public())
+        .unwrap();
+
+    // Verify reset on vault storage client with no namespace will clear all data
+    let mut storage = create_initialized_vault_storage(None);
+    storage
+        .reset_and_clear()
+        .expect("Failed to reset VaultStorage without namespace!");
+    assert!(storage_1.get(U64_KEY_1).is_err());
+    assert!(storage_2.get(U64_KEY_2).is_err());
+}
+
+/// Verifies that data stored and updated in distinct namespaces are entirely separate and
+/// independent.
+fn test_vault_distinct_namespaces() {
```

Comment:
This was tested already by the previous test in which we sent two different storage instances created at the same scope but with different namespaces. Not sure if this adds additional value.

Reply:
Unfortunately, the previous test that you're referring to (now called 'test_suite_multiple_namespaces()') doesn't actually test this functionality. While it does indeed run the test suite on different namespaces, it: (i) doesn't run tests concurrently, but rather sequentially against each namespace in turn; and (ii) calls reset after each individual test. As a result, it means that if namespaces don't actually provide isolation as we'd like, the test will still pass :( 

To avoid this, I've added this explicit test to ensure that isolation between namespaces is enforced, even in the case of concurrent and interleaved operations. This entire thing could be avoided should we support concurrent tests for vault.. but for now, it seems like a lot of work to try and make that a reality. If you still don't like the idea of this test being here, we can remove it, but I feel it does add some value. Thanks!

Reply:
This is appears to be a subset of this test: test_suite_multiple_namespaces

If this assumption is correct.... considering that we are going to be adding in something similar for other frameworks (namespaces), I'm not sure we benefit by having this special test. I'd prefer not to have it, because once we actually make tokens, policies, and namespaces a top level primitive, we should be able to merge the vault code into a generic test suite and we'll be wondering if we keep these tests or not.

What I'd like to see is a dead simple policy engine / tokens for in memory and on disk.

Reply:
Thanks for your response, David! While I still think I can present a counter example that shows this not to be the case (i.e., that test_suite_multiple_namespaces doesn't actually check for concurrent accesses across namespaces), I generally agree that this test will ultimately be removed once namespaces become a top-level primitive. As such, I'll remove this test for the time being and we can ensure the correct functionality is tested later on.